test/parallel/test-stream-readable-async-iterators.js

```diff
@@ -284,28 +280,6 @@ async function tests() {
     assert.strictEqual(received, 1);
   }
 
-  {
-    // Iterator throw.
-
-    const readable = new Readable({
-      objectMode: true,
-      read() {
-        this.push('hello');
-      }
-    });
-
-    readable.on('error', common.mustCall((err) => {
-      assert.strictEqual(err.message, 'kaboom');
-    }));
-
-    const it = readable[Symbol.asyncIterator]();
-    it.throw(new Error('kaboom')).catch(common.mustCall((err) => {
-      assert.strictEqual(err.message, 'kaboom');
-    }));
-
-    assert.strictEqual(readable.destroyed, true);
-  }
```

Comment:
@benjamingr are you ok in removing this? I don't remember who asked for this to be added in.

Reply:
I was the one that added it. It's a little unclear what and when exactly `throw` is used. Seems like async generator doesn't implement it. So I guess it's not important.

Reply:
`throw` throws an error inside an async generator:

```js
async function* asyncGen() {
  try { 
    yield;
  } catch (e) {
    console.log('got here', e);
  }
}
let gen = asyncGen();
gen.next();
gen.throw(new Error('kaboom'));
```

So the reason `throw` is used is to signal an error _in the opposite direction_ of the `error` event. So `return` is like attempting to close a stream (a generator may "resist" a return in the finally block) and a `throw` is like attempting to signal an error back to the stream when it yields.

It's OK for us to decide on whatever semantics we want, `for await` doesn't currently pass these values anyway and nothing in readable streams mimics this behavior.

As a fun fact, generators were designed to be "dual" like that with all the methods (return throw and next) taking values. So it's not only possible to model _readable_ streams with async generators but also _writeable_ streams (but not duplex streams):

```js
async function* writeable() {
  const target = [];
  try { 
    while(true) { target.push(yield); };
  } finally {
    console.log('got', target);
  }
}
const writeToMe = writeable();
writeToMe.next(); // start 
writeToMe.next('hello');
writeToMe.next('world');
writeToMe.return(); // causes the finally to run and print
```

Reply:
As a fun fact, you can also use _regular_ generators to model writeable streams in the above case, the "async" part could be the promise yielded signaling backpressure (for example reaching when writableHighWaterMark and then settling the promise when the stream can process data again).

Reply:
So we could do a `Writable.from` based on generators as well :D?

Reply:
Yes, although I see a ton of real world usage and value out of `Readable.from` and supporting async iterations on readable streams and almost no value in `Writeable.from` outside of "purity".

You will notice generators and async generators are quirky. `.return` has semantics that were tried but not expanded on in the language (where it runs finally blocks but not `catch blocks, the so called "third state"). `.throw` and `.return` look almost like they are built to make it easy to build async/await on top of generators. A proposal (compositional functions) tried to expand on this (desugaring async/await) but never got in.

Basically, `.throw` looks this way so what when implementing async/await you could throw an error if the promise you are yielding is rejected. At some point observables were supposed to be the "other side" with a duality proposal capitalizing on that fact. That also stalled :]

So basically I'm +0 on whatever we do here for `throw` (although I would probably expect it to destroy the stream?), it's mostly only important `.return` works correctly (since that's what breaking from inside a for await does).

I am -0 on Writeable.from since people _probably_ wouldn't use it? I had this debate with ry at some point regarding the I/O in deno (his main argument was that `.read` and `.write` are a lot more explicit than .next which I tend to agree with)

Reply:
I think we can resolve this conv?

Reply:
Sure though - probably uncomment the test? (if you add a `.next()` right after the start that "pumps" the async generator to the first `yield` at which point `throw`ing should indeed reach the catch and finally and destroy the stream).

Reply:
> if you add a .next() right after the start that "pumps" the async generator to the first yield

You lost me there

Reply:
I'm unable to make this test past. :/

Reply:
Want me to try and push to the branch?

Reply:
please do

Reply:
I think this is the main thing that we would need to solve.

Reply:
I'm unsure. I would guess that this is no relevant based on my understanding of @benjamingr's description.

Reply:
ok

Reply:
@benjamingr Please correct me if I'm wrong?

Reply:
Ok, we can resolve for me.